File: basic_dsa_three_hundred_forty_six.java
/**
 * Basic DSA Problems - Moving Average from Data Stream (LeetCode 346)
 * Calculate moving average over sliding window of size
 */
public class basic_dsa_three_hundred_forty_six {

    // Problem: Window size. next(val) adds val, returns avg of last size values.

    static class MovingAverage {
        final int size;
        final java.util.Queue<Integer> q = new java.util.LinkedList<>();
        int sum = 0;

        MovingAverage(int size) { this.size = size; }

        public double next(int val) {
            q.offer(val);
            sum += val;
            if (q.size() > size) sum -= q.poll();
            return (double) sum / q.size();
        }
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Moving Average from Data Stream");
        System.out.println("=".repeat(60));

        MovingAverage ma = new MovingAverage(3);
        System.out.println("next(1)=" + ma.next(1) + ", next(10)=" + ma.next(10)
                + ", next(3)=" + ma.next(3) + ", next(5)=" + ma.next(5));
    }
}

File: basic_dsa_three_hundred_forty_six.cpp
/**
 * Basic DSA - N-th Tribonacci Number (LeetCode 1137)
 * T(0)=0, T(1)=T(2)=1, T(n)=T(n-1)+T(n-2)+T(n-3).
 */
#include <iostream>
#include <string>

int tribonacci(int n) {
    if (n == 0) return 0;
    if (n <= 2) return 1;
    int a = 0, b = 1, c = 1;
    for (int i = 3; i <= n; i++) {
        int d = a + b + c;
        a = b;
        b = c;
        c = d;
    }
    return c;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: N-th Tribonacci Number\n";
    std::cout << std::string(60, '=') << "\n";

    std::cout << "4 -> " << tribonacci(4) << "\n";
    std::cout << "25 -> " << tribonacci(25) << "\n";
}
